### ISSUE TITLE
[Question]  When should I pass "Transformers" to the method `model.predict_on_batch` or `model.predict`?

### ISSUE THREAD

**Issue by mohammad-saber** [deepchem/deepchem#2662]
## ❓ Questions & Help

Hi, Thank you for sharing your great code. I just started to use DeepChem. 

I am following the [tutorial](https://github.com/deepchem/deepchem/blob/master/examples/tutorials/01_The_Basic_Tools_of_the_Deep_Life_Sciences.ipynb). I have a few questions about `model.predict_on_batch` and `transformers`:

Q1) According to the [documentation](https://deepchem.readthedocs.io/en/latest/api_reference/models.html#deepchem.models.KerasModel.predict_on_batch), `predict_on_batch` has an argument called `transformers`. However, in the following line of the code, `transformers` is not passed. My question is when I should pass `transformers` to the method `model.predict_on_batch` or `model.predict`.

`solubilities = model.predict_on_batch(test_dataset.X[:10])
`

Q2) According to the [documentation](https://deepchem.readthedocs.io/en/latest/api_reference/models.html#deepchem.models.KerasModel.predict_on_batch), `transformers` is:

`Transformers that the input data has been transformed by. The output is passed through these transformers to undo the transformations.`

I have misunderstanding about `transformers`.

Q2-1) In the above definition, what does `input data` mean? Is it X data (features or input to the model) or Y data (model output)?

Q2-2) Why do we need to undo transformation on output?

From my previous experience in machine learning, I used to apply Normalization or Standard Scaling on X (input to the model or features). I would like to ask the reason of `output is passed through these transformers to undo the transformations`.

Q3) What is the different between `model.predict_on_batch` or `model.predict`? If I want to get prediction for the whole test dataset, should I use:

`solubilities = model.predict(test_dataset.X)
`

Thank you for your time to read my questions.

**Comment by arunppsg** [comment]
Hey, sorry for a very late response. From the best of my understanding,

On 1)
Transformer has to be used on test dataset when the train dataset is transformed using a transformer for training. In such cases, `transformers` parameters can be used to specify the same transformation used on the training dataset to the test dataset. When a dataset loaded from `dc.molnet` submodule is used, `transformers` argument need not be passed during prediction because the train dataset and test datasets are already transformed (transformer argument is passed during loading as an argument to `dc.molnet.load_X()`).

On 2-1) `input data` refers to train dataset. The second line - `The output is passed through these transformers to undo the transformations` is not clear for me also. Tagging @rbharath for this.

On 3) 
`model.predict_on_batch` predicts on a sample of data. It can be used like `model.predict_on_batch(test_dataset.X)` or it can also be used to predict on a set of samples like `model.predict_on_batch(test_dataset.X[:10])`.  
`model.predict` is used to predict on the entire dataset. It takes `deepchem.data.datasets.Dataset` as parameter. Example usage: `model.predict(test_dataset)`.

**Comment by mohammad-saber** [comment]
Thank you for your help. 

I guess the meaning of `The output is passed through these transformers to undo the transformations` is :

We want to get model output in original scale (without transformation) for the purpose of plotting or comparing with `y_true` in original scale.

**Comment by arunppsg** [comment]
Yes, when the model is making predictions, the predictions will be transformed outputs. Hence, we are undoing the transformation applied on the predictions to get the original value.

A bit of example: The true y is of range 0-100, suppose we transform it to a scale of 0-1 during training, the model output will also be in the scale of 0-1. Hence, we undo the transformation in the output by passing it to the transformers to get the predicted  output in the original scale. For generating loss function, we use the scaled values since it gives more numerical stability.

**Comment by mohammad-saber** [comment]
Thank you for your help.
